In [106]:
import findspark
findspark.init()
from pyspark.sql import SparkSession
spark = SparkSession.builder \
.master("local[4]") \
.appName("spark_machine") \
.getOrCreate()

In [107]:
df = spark.read\
.option("header", True) \
.option("inferSchema","True")\
.option("sep",",")\
.csv("C://Users//oguzh//OneDrive//Masaüstü//adult.data")
df.show(5)

+---+-----------------+--------+----------+-------------+-------------------+------------------+--------------+------+-------+------------+------------+--------------+--------------+------+
|age|        workclass|  fnlwgt| education|education_num|     marital_status|        occupation|  relationship|  race|    sex|capital_gain|capital_loss|hours_per_week|native_country|output|
+---+-----------------+--------+----------+-------------+-------------------+------------------+--------------+------+-------+------------+------------+--------------+--------------+------+
| 39|        State-gov| 77516.0| Bachelors|         13.0|      Never-married|      Adm-clerical| Not-in-family| White|   Male|      2174.0|         0.0|          40.0| United-States| <=50K|
| 50| Self-emp-not-inc| 83311.0| Bachelors|         13.0| Married-civ-spouse|   Exec-managerial|       Husband| White|   Male|         0.0|         0.0|          13.0| United-States| <=50K|
| 38|          Private|215646.0|   HS-grad|       

In [108]:
df_test = spark.read\
.option("header", True) \
.option("inferSchema","True")\
.option("sep",",")\
.csv("C://Users//oguzh//OneDrive//Masaüstü//adult.test")
df.show(5)

+---+-----------------+--------+----------+-------------+-------------------+------------------+--------------+------+-------+------------+------------+--------------+--------------+------+
|age|        workclass|  fnlwgt| education|education_num|     marital_status|        occupation|  relationship|  race|    sex|capital_gain|capital_loss|hours_per_week|native_country|output|
+---+-----------------+--------+----------+-------------+-------------------+------------------+--------------+------+-------+------------+------------+--------------+--------------+------+
| 39|        State-gov| 77516.0| Bachelors|         13.0|      Never-married|      Adm-clerical| Not-in-family| White|   Male|      2174.0|         0.0|          40.0| United-States| <=50K|
| 50| Self-emp-not-inc| 83311.0| Bachelors|         13.0| Married-civ-spouse|   Exec-managerial|       Husband| White|   Male|         0.0|         0.0|          13.0| United-States| <=50K|
| 38|          Private|215646.0|   HS-grad|       

# DF çok geniş olduğu için düzgün bir görünüm alamadık. Driver'a aldığımız birkaç satırdan oluşan dataframe'i pandas dataframe'e çevirerek daha güzel bir görünüm sağlayabiliriz


In [109]:
df.limit(5).toPandas().head()

age          workclass    fnlwgt   education  education_num  \
0   39          State-gov   77516.0   Bachelors           13.0   
1   50   Self-emp-not-inc   83311.0   Bachelors           13.0   
2   38            Private  215646.0     HS-grad            9.0   
3   53            Private  234721.0        11th            7.0   
4   28            Private  338409.0   Bachelors           13.0   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  output  
0        2174.0           0.0            40.0   United-States   <=50K  
1           0.0           0.0            13.0   United-States   <=50K  
2           0.0           0.0            40.0   United-States   <=50K  
3           0.0           0.0            40.0   United-States   <=50K  
4           0.0           0.0            40.0            Cuba   <=50K

In [110]:
df_test.limit(5).toPandas().head()

age   workclass    fnlwgt      education  education-num  \
0   25     Private  226802.0           11th            7.0   
1   38     Private   89814.0        HS-grad            9.0   
2   28   Local-gov  336951.0     Assoc-acdm           12.0   
3   44     Private  160323.0   Some-college           10.0   
4   18           ?  103497.0   Some-college           10.0   

        marital-status          occupation relationship    race      sex  \
0        Never-married   Machine-op-inspct    Own-child   Black     Male   
1   Married-civ-spouse     Farming-fishing      Husband   White     Male   
2   Married-civ-spouse     Protective-serv      Husband   White     Male   
3   Married-civ-spouse   Machine-op-inspct      Husband   Black     Male   
4        Never-married                   ?    Own-child   White   Female   

   capital-gain  capital-loss  hours-per-week  native-country   output  
0           0.0           0.0            40.0   United-States   <=50K.  
1           0.0           0.0            50.0   United-States   <=50K.  
2           0.0           0.0            40.0   United-States    >50K.  
3        7688.0           0.0            40.0   United-States    >50K.  
4           0.0           0.0            30.0   United-States   <=50K.

In [111]:
last_df = df.union(df_test)
last_df.limit(5).toPandas().head()

age          workclass    fnlwgt   education  education_num  \
0   39          State-gov   77516.0   Bachelors           13.0   
1   50   Self-emp-not-inc   83311.0   Bachelors           13.0   
2   38            Private  215646.0     HS-grad            9.0   
3   53            Private  234721.0        11th            7.0   
4   28            Private  338409.0   Bachelors           13.0   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  output  
0        2174.0           0.0            40.0   United-States   <=50K  
1           0.0           0.0            13.0   United-States   <=50K  
2           0.0           0.0            40.0   United-States   <=50K  
3           0.0           0.0            40.0   United-States   <=50K  
4           0.0           0.0            40.0            Cuba   <=50K

In [112]:
last_df.printSchema()

root
 |-- age: integer (nullable = true)
 |-- workclass: string (nullable = true)
 |-- fnlwgt: double (nullable = true)
 |-- education: string (nullable = true)
 |-- education_num: double (nullable = true)
 |-- marital_status: string (nullable = true)
 |-- occupation: string (nullable = true)
 |-- relationship: string (nullable = true)
 |-- race: string (nullable = true)
 |-- sex: string (nullable = true)
 |-- capital_gain: double (nullable = true)
 |-- capital_loss: double (nullable = true)
 |-- hours_per_week: double (nullable = true)
 |-- native_country: string (nullable = true)
 |-- output: string (nullable = true)



# numerik değişkenleri analiz ettim.

In [113]:
last_df.describe(["age","fnlwgt","education_num","capital_gain","capital_loss","hours_per_week"]).toPandas().head()

summary                 age              fnlwgt       education_num  \
0   count               48842               48842               48842   
1    mean   38.64358543876172  189664.13459727284  10.078088530363212   
2  stddev  13.710509934443563  105604.02542315733   2.570972755592263   
3     min                  17             12285.0                 1.0   
4     max                  90           1490400.0                16.0   

         capital_gain       capital_loss      hours_per_week  
0               48842              48842               48842  
1  1079.0676262233324  87.50231358257237  40.422382375824085  
2   7452.019057655401  403.0045521243599    12.3914440242523  
3                 0.0                0.0                 1.0  
4             99999.0             4356.0                99.0

# Kategorik değişkenler için bir fonksiyon yazıp groupby ile bilgi almaya çalıştım.

In [114]:
def freq_table(df, cols):
    return df.groupBy(cols).count().toPandas().head(10)

In [115]:
freq_table(last_df, ["workclass"])

workclass  count
0          State-gov   1981
1        Federal-gov   1432
2   Self-emp-not-inc   3862
3          Local-gov   3136
4            Private  33906
5                  ?   2799
6       Self-emp-inc   1695
7        Without-pay     21
8       Never-worked     10

2.799 adet ? var.Bunlar büyük ihitmal kayıp bilgi 
Daha sonra never-worked ve without-pay sınıfları çok az tekrarlanmış. Bunların da veri setinden çıkarılması düşünebilir. Yada birleştirilip genelleme yapmak düşünülebilir, ama risklidir farklı sınıfları birleştirirken sonucu etkilemesine göre karar verilir.

In [116]:
freq_table(last_df, ["education"])

education  count
0   Prof-school    834
1          10th   1389
2       7th-8th    955
3       5th-6th    509
4    Assoc-acdm   1601
5     Assoc-voc   2061
6       Masters   2657
7          12th    657
8     Preschool     83
9           9th    756

Genel olarak bir sıkıntı görünmüyor gibi, ama çok fazla kategori var ben sınırlama yaptığımdan görünmüyor. Bu kategoriler birleştirilebilir. İlk - orta - lise - üniversite - yüksek_lisans gibi genelleme yapılabilir.

In [117]:
freq_table(last_df, ["marital_status"])

marital_status  count
0                 Widowed   1518
1   Married-spouse-absent    628
2       Married-AF-spouse     37
3      Married-civ-spouse  22379
4                Divorced   6633
5           Never-married  16117
6               Separated   1530

Bir sorun görünmüyor ama ayrılmış ile boşanmış aynı anlama geliyor gibi görünüyor. Diğer parametrelere bakılarak birleştirme işlemleri denenebilir.

In [118]:
freq_table(last_df, ["occupation"])

occupation  count
0     Farming-fishing   1490
1   Handlers-cleaners   2072
2      Prof-specialty   6172
3        Adm-clerical   5611
4     Exec-managerial   6086
5        Craft-repair   6112
6               Sales   5504
7                   ?   2809
8        Tech-support   1446
9    Transport-moving   2355

Bakıldığında sorun yokmuş gibi duruyor lakin ? 2809 tane veri var bunlar diğer sınıfların verisini içeriyor olabilir. Bunun analizi yapılmalı 

In [119]:
freq_table(last_df, ["relationship"])

relationship  count
0          Husband  19716
1        Own-child   7581
2    Not-in-family  12583
3   Other-relative   1506
4             Wife   2331
5        Unmarried   5125

Bir sorun yok gibi

In [120]:
freq_table(last_df, ["race"])

race  count
0   Asian-Pac-Islander   1519
1                Black   4685
2                Other    406
3                White  41762
4   Amer-Indian-Eskimo    470

Bir sıkıntı yok beklenilen gibi beyaz ve siyahlarda çoğunluk görülmekte.

In [121]:
freq_table(last_df, ["sex"])

sex  count
0     Male  32650
1   Female  16192

Erkeklerde fazlalık var bir sorun yok.

In [122]:
freq_table(last_df, ["native_country"])

native_country  count
0   Dominican-Republic    103
1              Ireland     37
2                 Cuba    138
3            Guatemala     88
4                 Iran     59
5               Taiwan     65
6          El-Salvador    155
7        United-States  43832
8                South    115
9                Japan     92

In [123]:
freq_table(last_df, ["output"])

output  count
0     >50K   7841
1    >50K.   3846
2   <=50K.  12435
3    <=50K  24720

"." içeren veriler var bunlar düzeltilmeli

In [124]:
last_df.toPandas().head()

Exception ignored in: <object repr() failed>
Traceback (most recent call last):
  File "C:\spark\python\pyspark\ml\wrapper.py", line 105, in __del__
    SparkContext._active_spark_context._gateway.detach(self._java_obj)
AttributeError: 'OneHotEncoderEstimator' object has no attribute '_java_obj'


age          workclass    fnlwgt   education  education_num  \
0   39          State-gov   77516.0   Bachelors           13.0   
1   50   Self-emp-not-inc   83311.0   Bachelors           13.0   
2   38            Private  215646.0     HS-grad            9.0   
3   53            Private  234721.0        11th            7.0   
4   28            Private  338409.0   Bachelors           13.0   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  output  
0        2174.0           0.0            40.0   United-States   <=50K  
1           0.0           0.0            13.0   United-States   <=50K  
2           0.0           0.0            40.0   United-States   <=50K  
3           0.0           0.0            40.0   United-States   <=50K  
4           0.0           0.0            40.0            Cuba   <=50K

In [125]:
from pyspark.sql.functions import *

# Veri Temizliği 

Tüm kategorik sütunlara boşluk kontrolu yaptım.

In [126]:
kategorik = ["workclass", "education", "marital_status","occupation","relationship","race","sex","native_country","output"]
for degisken in kategorik:
    df2 = last_df.withColumn(degisken, trim(last_df[degisken]))

In [127]:
df2.count()

48842

In [128]:
last_df.count()

48842

# output değişkeni içindeki noktaları temizleyelim

In [129]:
df3 = df2 \
.withColumn("output",regexp_replace(col("output"),"<=50K.","<=50K")) \
.withColumn("output",regexp_replace(col("output"),">50K.",">50K"))

In [130]:
df3.groupby(df3["output"]).agg({"*":"count"}) \
.toPandas().head()

output  count(1)
0  <=50K     37155
1   >50K     11687

# null kontrolü yapalım

In [131]:
for degisken in df3.columns:
  bos = df3.filter(df3[degisken].isNull()).count()
  print(degisken + " değişkeninde " + str(bos) + " adet boş veri var.")

age değişkeninde 0 adet boş veri var.
workclass değişkeninde 0 adet boş veri var.
fnlwgt değişkeninde 0 adet boş veri var.
education değişkeninde 0 adet boş veri var.
education_num değişkeninde 0 adet boş veri var.
marital_status değişkeninde 0 adet boş veri var.
occupation değişkeninde 0 adet boş veri var.
relationship değişkeninde 0 adet boş veri var.
race değişkeninde 0 adet boş veri var.
sex değişkeninde 0 adet boş veri var.
capital_gain değişkeninde 0 adet boş veri var.
capital_loss değişkeninde 0 adet boş veri var.
hours_per_week değişkeninde 0 adet boş veri var.
native_country değişkeninde 0 adet boş veri var.
output değişkeninde 0 adet boş veri var.


# ? işaretleri ile ilgili problemlerimiz vardı şimdi onlarla ilgilenelim.

In [132]:
for degisken in df3.columns:
  deger = df3.filter(df3[degisken].contains("?")).count()
  print(degisken + " değişkeninde " + str(deger) + " adet ?'li veri var.")

age değişkeninde 0 adet ?'li veri var.
workclass değişkeninde 2799 adet ?'li veri var.
fnlwgt değişkeninde 0 adet ?'li veri var.
education değişkeninde 0 adet ?'li veri var.
education_num değişkeninde 0 adet ?'li veri var.
marital_status değişkeninde 0 adet ?'li veri var.
occupation değişkeninde 2809 adet ?'li veri var.
relationship değişkeninde 0 adet ?'li veri var.
race değişkeninde 0 adet ?'li veri var.
sex değişkeninde 0 adet ?'li veri var.
capital_gain değişkeninde 0 adet ?'li veri var.
capital_loss değişkeninde 0 adet ?'li veri var.
hours_per_week değişkeninde 0 adet ?'li veri var.
native_country değişkeninde 857 adet ?'li veri var.
output değişkeninde 0 adet ?'li veri var.


###  ? olan değişkenlerin output değişkeniyle olan ilişkisine bir bakalım

In [133]:
df3.select("workclass","occupation","native_country") \
.filter(
df3["workclass"].contains("?") |
df3["occupation"].contains("?") |
df3["native_country"].contains("?")
) \
.groupBy("workclass","occupation","native_country").count()\
.orderBy(col("count").desc())\
.toPandas().head(35)

workclass          occupation  native_country  count
0                   ?                   ?   United-States   2530
1             Private       Other-service               ?    106
2             Private      Prof-specialty               ?     99
3             Private        Craft-repair               ?     66
4             Private               Sales               ?     65
5             Private     Exec-managerial               ?     61
6             Private        Adm-clerical               ?     57
7                   ?                   ?          Mexico     48
8             Private   Machine-op-inspct               ?     46
9                   ?                   ?               ?     46
10            Private    Transport-moving               ?     31
11            Private   Handlers-cleaners               ?     24
12            Private        Tech-support               ?     22
13                  ?                   ?          Canada     19
14          State-gov      Prof-specialty               ?     18
15          Local-gov      Prof-specialty               ?     17
16   Self-emp-not-inc               Sales               ?     17
17       Self-emp-inc     Exec-managerial               ?     17
18                  ?                   ?           South     14
19        Federal-gov      Prof-specialty               ?     14
20   Self-emp-not-inc      Prof-specialty               ?     13
21       Self-emp-inc               Sales               ?     13
22   Self-emp-not-inc     Exec-managerial               ?     13
23                  ?                   ?         Germany     13
24                  ?                   ?     Philippines     12
25   Self-emp-not-inc        Craft-repair               ?     12
26       Never-worked                   ?   United-States     10
27            Private     Priv-house-serv               ?     10
28                  ?                   ?          Taiwan     10
29       Self-emp-inc        Craft-repair               ?      9
30                  ?                   ?     Puerto-Rico      9
31                  ?                   ?           China      9
32                  ?                   ?     El-Salvador      8
33                  ?                   ?         England      8
34            Private     Farming-fishing               ?      7

### Soru işaretlerinin dağılımına baktığımızda hedef değişkenle arasında tesadüfi bir ilişki bulunuyor silmek mantıklı olacağını düşündüm.

In [134]:
df4 = df3\
.filter(~(
    df3["workclass"].contains("?") |
    df3["occupation"].contains("?") |
    df3["native_country"].contains("?")
))
print(df3.count())
print(df4.count())

48842
45222


### Zayıf Sınıfların Kaldırılması workclass niteliğinde never-worked ve without-pay sınıfları ve occupation niteliğinde Armed-Forces sınıfı çok az tekrarlanmış. Veri setinden çıkartabiliriz.

In [135]:
df5 = df4 \
.filter(~(df4["workclass"].contains("never-worked") |
        df4["workclass"].contains("without-pay") |
        df4["occupation"].contains("Armed-Forces") |
        df4["native_country"].contains("Holand-Netherlands") 
         ))
print(df4.count())
print(df5.count())

45222
45207


### Eğitim durumlarıyla ilgili bir düzenleme yapalım.

######         1st-4th , 5th-6th , 7th-8th : elementary-school
######         9th,10th,11th,12th : high school
######          Masters,Doctorate:Postgraduate
######          Bachelors, Some-college : undergraduate
######          bu şekilde bir birleştirme yapalım.

In [136]:
df5 = df5 \
.withColumn("workclass", trim(col("workclass"))) \
.withColumn("education", trim(col("education"))) \
.withColumn("marital_status", trim(col("marital_status"))) \
.withColumn("occupation", trim(col("occupation"))) \
.withColumn("relationship", trim(col("relationship"))) \
.withColumn("race", trim(col("race"))) \
.withColumn("sex", trim(col("sex"))) \
.withColumn("native_country", trim(col("native_country"))) \
.withColumn("output", trim(col("output")))

In [137]:
df6 = df5.withColumn("education_merged", \
      when(col("education").isin("1st-4th","education","5th-6th","education","7th-8th"), "Elementary-School")
        .when(col("education").isin("9th","10th","11th","12th"), "High-School")
        .when(col("education").isin("Masters","Doctorate"), "Post-Graduate")
        .when(col("education").isin("Bachelors","Some-college"), "Under-Graduate")
        .otherwise(col("education")))
df6.select("education","education_merged").toPandas().head(20)

education   education_merged
0      Bachelors     Under-Graduate
1      Bachelors     Under-Graduate
2        HS-grad            HS-grad
3           11th        High-School
4      Bachelors     Under-Graduate
5        Masters      Post-Graduate
6            9th        High-School
7        HS-grad            HS-grad
8        Masters      Post-Graduate
9      Bachelors     Under-Graduate
10  Some-college     Under-Graduate
11     Bachelors     Under-Graduate
12     Bachelors     Under-Graduate
13    Assoc-acdm         Assoc-acdm
14       7th-8th  Elementary-School
15       HS-grad            HS-grad
16       HS-grad            HS-grad
17          11th        High-School
18       Masters      Post-Graduate
19     Doctorate      Post-Graduate

###### nitelikleri önce kategorik sonra numerik şeklinde ayarladım.

In [138]:
nitelik_siralama = ["workclass", "education", "education_merged", "marital_status", "occupation", "relationship", "race",
     "sex", "native_country", "age", "fnlwgt", "education_num", "capital_gain", "capital_loss", "hours_per_week","output"]
df7 = df6.select(nitelik_siralama)
df7.toPandas().head()

workclass  education education_merged      marital_status  \
0         State-gov  Bachelors   Under-Graduate       Never-married   
1  Self-emp-not-inc  Bachelors   Under-Graduate  Married-civ-spouse   
2           Private    HS-grad          HS-grad            Divorced   
3           Private       11th      High-School  Married-civ-spouse   
4           Private  Bachelors   Under-Graduate  Married-civ-spouse   

          occupation   relationship   race     sex native_country  age  \
0       Adm-clerical  Not-in-family  White    Male  United-States   39   
1    Exec-managerial        Husband  White    Male  United-States   50   
2  Handlers-cleaners  Not-in-family  White    Male  United-States   38   
3  Handlers-cleaners        Husband  Black    Male  United-States   53   
4     Prof-specialty           Wife  Black  Female           Cuba   28   

     fnlwgt  education_num  capital_gain  capital_loss  hours_per_week output  
0   77516.0           13.0        2174.0           0.0            40.0  <=50K  
1   83311.0           13.0           0.0           0.0            13.0  <=50K  
2  215646.0            9.0           0.0           0.0            40.0  <=50K  
3  234721.0            7.0           0.0           0.0            40.0  <=50K  
4  338409.0           13.0           0.0           0.0            40.0  <=50K

In [139]:
df8 = df7.drop("education")
df8.toPandas().head()

workclass education_merged      marital_status         occupation  \
0         State-gov   Under-Graduate       Never-married       Adm-clerical   
1  Self-emp-not-inc   Under-Graduate  Married-civ-spouse    Exec-managerial   
2           Private          HS-grad            Divorced  Handlers-cleaners   
3           Private      High-School  Married-civ-spouse  Handlers-cleaners   
4           Private   Under-Graduate  Married-civ-spouse     Prof-specialty   

    relationship   race     sex native_country  age    fnlwgt  education_num  \
0  Not-in-family  White    Male  United-States   39   77516.0           13.0   
1        Husband  White    Male  United-States   50   83311.0           13.0   
2  Not-in-family  White    Male  United-States   38  215646.0            9.0   
3        Husband  Black    Male  United-States   53  234721.0            7.0   
4           Wife  Black  Female           Cuba   28  338409.0           13.0   

   capital_gain  capital_loss  hours_per_week output  
0        2174.0           0.0            40.0  <=50K  
1           0.0           0.0            13.0  <=50K  
2           0.0           0.0            40.0  <=50K  
3           0.0           0.0            40.0  <=50K  
4           0.0           0.0            40.0  <=50K

In [164]:
num_cols = ["age","fnlwgt","education_num","capital_gain","capital_loss","hours_per_week"]
label = ["output"]
cat_cols = ["workclass","education_merged","marital_status","occupation","relationship","race","sex","native_country"]

##### Gerekli kütüphaneleri ekledik 

In [165]:
from pyspark.ml.feature import StringIndexer, OneHotEncoderEstimator, VectorAssembler
from pyspark.ml import Pipeline, PipelineModel

###### spark sürümünden kaynaklı olarak aşağıdaki gibi bir kod yazdım. Yeni sürümlerde böyle değil çok daha basit şekilde tek satırda halledebiliyoruz. Biraz araştırdım ama bu sürümde yapamadım.

In [166]:
workclass_indexer = StringIndexer() \
.setInputCol("workclass") \
.setOutputCol("workclass_index") \
.setHandleInvalid("skip")
education_merged_indexer = StringIndexer() \
.setInputCol("education_merged") \
.setOutputCol("education_merged_index") \
.setHandleInvalid("skip")
marital_status_indexer = StringIndexer() \
.setInputCol("marital_status") \
.setOutputCol("marital_status_index") \
.setHandleInvalid("skip")
occupation_indexer = StringIndexer() \
.setInputCol("occupation") \
.setOutputCol("occupation_index") \
.setHandleInvalid("skip")
relationship_indexer = StringIndexer() \
.setInputCol("relationship") \
.setOutputCol("relationship_index") \
.setHandleInvalid("skip")
race_indexer = StringIndexer() \
.setInputCol("race") \
.setOutputCol("race_index") \
.setHandleInvalid("skip")
native_country_indexer = StringIndexer() \
.setInputCol("native_country") \
.setOutputCol("native_country_index") \
.setHandleInvalid("skip")

encoder = OneHotEncoderEstimator() \
.setInputCols(["workclass_index","education_merged_index","marital_status_index","occupation_index","relationship_index",
               "race_index", "native_country_index"]) \
.setOutputCols(["workclass_index_enc","education_merged_index_enc","marital_status_index_enc","occupation_index_enc",
                "relationship_index_enc", "race_index_enc", "native_country_index_enc"])

###### VectorAssembler, label_indexer, StandardScale işlemlerini yaptık.

In [167]:
vector_assembler = VectorAssembler().setInputCols(num_cols + encoder.getOutputCols()).setOutputCol("vectorized_features")

In [168]:
label_indexer = StringIndexer() \
.setInputCol("output") \
.setOutputCol("label")

In [169]:
from pyspark.ml.feature import StandardScaler 
scaler = StandardScaler(inputCol="vectorized_features",outputCol="features", withStd=True, withMean=True)

###### basit bir uygulama olduğu için logisticregression algoritmasını kullandım.

In [170]:
from pyspark.ml.classification import LogisticRegression

In [171]:
lr_object = LogisticRegression() \
.setFeaturesCol("features") \
.setLabelCol("label") \
.setPredictionCol("prediction")

In [172]:
train_df, test_df = df8.randomSplit([0.8, 0.2], seed=142)

###### pipeline oluşturup modelimi eğitip test ettim.

In [173]:
pipeline_nesnesi = Pipeline() \
.setStages([workclass_indexer, education_merged_indexer, marital_status_indexer, occupation_indexer, relationship_indexer, 
            race_indexer, native_country_indexer, encoder, vector_assembler, label_indexer, scaler, lr_object])

In [174]:
pipeline_model = pipeline_nesnesi.fit(train_df)

In [175]:
sonuc_df = pipeline_model.transform(test_df).select("label","prediction")
sonuc_df.toPandas().head()

label  prediction
0    0.0         0.0
1    0.0         0.0
2    1.0         0.0
3    0.0         0.0
4    1.0         1.0

In [186]:
from sklearn.metrics import accuracy_score
y_pred = sonuc_df.select("prediction").toPandas()
y_true = sonuc_df.select("label").toPandas()

In [187]:
accuracy_score(y_true, y_pred)

0.8461453623028565

###### Basit bir uygulama yapmış olduk. Apache Spark ile bir sınıflandırma problemi tamamladık. Teşekkürler